File: UCFPT/22locals/local2022FinalRoundData/solutions/whichnum/whichnum.py

import sys
input = sys.stdin.readline

# Check if a number is prime
def primeCheck(a):
    assert a > 1
    i = 2
    while i * i < a:
        assert a % i
        i += 1
            

# Read in the header line
line = input()
n, k = map(int, line.strip().split())
assert str(n) + " " + str(k) + "\n" == line
assert 1 <= n <= 10**17
assert 1 <= k <= 14

# Read in the values
line = input()
arr = list(map(int, line.strip().split()))
assert " ".join(str(x) for x in arr) + "\n" == line
product = 1
for x in arr:
    primeCheck(x)
    product *= x
assert 1 <= product <= 10**17

# Make the array of the divisor the inclusion exclusion
masks = [0] * (1<<k)
masks[0] = 1
for i in range(k):
    masks[1<<i] = -arr[i]
for i in range(1<<k):
    if masks[i]:
        continue
    masks[i] = masks[i-(i&-i)] * masks[i&-i]

# Use inclusion exclusion to find how many numbers are spoken up to some given value
def pos(value):
    global arr,masks
    safe = 1
    for x in arr:
        if (value%x) == 0:
            safe = 0
    ans = sum((value // abs(x))*(1 if x > 0 else -1) for x in masks)
    return ans, safe

# Find the initial bounds for our binary search
lo = 0
# Actual high bound should be less than 5*10**17
hi = 10**19

# Binary search for the value by counting how many numbers are spoken up to the given point
while lo < hi:
    m = (lo + hi) // 2
    p,safe = pos(m)
    if p == n and safe:
        print(m)
        exit()
    if p < n:
        lo = m + 1
    else:
        hi = m
assert 0

File: UCFPT/22locals/local2022FinalRoundData/solutions/whichnum/whichnum.java
// Arup Guha
// 8/9/2022
// Solution to 2022 UCF Locals Problem: Which Number

import java.util.*;

public class whichnum {
	
	public static int numP;
	public static long[] pList;
	
	public static void main(String[] args) {
	
		// Read input.
		Scanner stdin = new Scanner(System.in);
		long n = stdin.nextLong();
		numP = stdin.nextInt();
		pList = new long[numP];
		for (int i=0; i<numP; i++)
			pList[i] = stdin.nextLong();
		
		// Set bounds for binary search.
		long low = 1, high = 1000000000000000000L;
		while (isDivisible(high)) high++;
		
		// Run binary search.
		while (low < high) {
		
			// Get next and previous valid number to try.
			long mid = (low+high)/2;
			long next = mid, prev = mid-1;
			while (isDivisible(next)) next++;
			while (isDivisible(prev)) prev--;
			
			// Get how many #s Natasha says here.
			long numValid = count(next);
			
			// Must go higher.
			if (numValid < n)
				low = next+1;
				
			// Must go lower and we know it can't be higher than prev.
			else if (numValid > n)
				high = prev;
				
			// We're good. This will let us break out.	
			else 
				low = high = next;
		}
		
		// Ta da!
		System.out.println(low);
	}
	
	public static long count(long x) {
	
		long res = x;
		
		// Store products for each mask subset.
		long[] prod = new long[1<<numP];
		prod[0] = 1;
		
		// Run inclusion-exclusion.
		for (int mask=1; mask<(1<<numP); mask++) {
		
			// Decide if we're adding or subtracting.
			int sign = Integer.bitCount(mask)%2 == 0 ? 1 : -1;
			
			// Lowest one bit.
			int bit = 0;
			while ((mask & (1<<bit)) == 0) bit++;
			
			// Update this product.
			prod[mask] = prod[mask-(1<<bit)]*pList[bit];
			
			// Number of terms in range divisible.
			long tmp = x/prod[mask];
			
			// Add or subtract as needed.
			res += (sign*tmp);
		}
		
		return res;
	}
	
	// Returns true iff x is divisible by one of the primes we are to avoid.
	public static boolean isDivisible(long x) {
		for (Long p : pList)
			if (x%p == 0)
				return true;
		return false;
	}
}